### DISCUSSION TITLE
How to work and build individual modules?

### DISCUSSION THREAD

**Discussion by moacirrf** [apache/netbeans#4664] [Q&A] [answered]
Hi I'm working on possible improvement of module Markdown Support, based on what i'm doing on plugin
https://github.com/moacirrf/netbeans-markdown
 
I can't compile the module, there is a way to compile a single module without build Netbeans entirely?

Thanks

> ant -f /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/ide/markdown -Dcontinue.after.failing.tests=true -Ddo.not.clean.module.config.xml=true clean build
init-tasks:
default.init:
Updating property file: /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/moduleCluster.properties
common-init:
basic-init:
projectized-common.-release.files:
projectized-common.files-init:
files-init:
testuserdir-delete:
projectized-common.clean:
clean:
Loading module list from /tmp/nb-scan-cache-809e9e85-full.ser
Cannot translate according to [file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.filesystems/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.loaders/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.nodes/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.text/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.util/module-auto-deps.xml] because could not find /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/lib/boot.jar
Cannot translate according to [file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.filesystems/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.loaders/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.nodes/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.text/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.util/module-auto-deps.xml] because could not find /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/lib/boot.jar
init-tasks:
default.init:
Updating property file: /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/moduleCluster.properties
common-init:
basic-init:
projectized-common.-release.files:
projectized-common.files-init:
files-init:
jdk-8-check:
Build number : 20220921-5c4f116a1a00
set-buildnumber:
build-init:
Cannot translate according to [file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.filesystems/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.loaders/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.nodes/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.text/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.util/module-auto-deps.xml] because could not find /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/lib/boot.jar
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/modules/org-netbeans-core-multiview.jar missing for /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/ide/markdown; will first try to build /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/core.multiview
platform/core.multiview.init-tasks:
platform/core.multiview.default.init:
Updating property file: /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/moduleCluster.properties
platform/core.multiview.common-init:
platform/core.multiview.basic-init:
platform/core.multiview.projectized-common.-release.files:
platform/core.multiview.projectized-common.files-init:
platform/core.multiview.files-init:
platform/core.multiview.jdk-8-check:
Build number : 20220921-5c4f116a1a00
platform/core.multiview.set-buildnumber:
platform/core.multiview.build-init:
Loading module list from /tmp/nb-scan-cache-809e9e85-full.ser
Cannot translate according to [file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.filesystems/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.nodes/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.text/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.util/module-auto-deps.xml] because could not find /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/lib/boot.jar
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/modules/org-openide-dialogs.jar missing for /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/core.multiview; will first try to build /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.dialogs
platform/openide.dialogs.init-tasks:
platform/openide.dialogs.default.init:
Updating property file: /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/moduleCluster.properties
platform/openide.dialogs.common-init:
platform/openide.dialogs.basic-init:
platform/openide.dialogs.projectized-common.-release.files:
platform/openide.dialogs.projectized-common.files-init:
platform/openide.dialogs.files-init:
platform/openide.dialogs.jdk-8-check:
Build number : 20220921-5c4f116a1a00
platform/openide.dialogs.set-buildnumber:
platform/openide.dialogs.build-init:
Loading module list from /tmp/nb-scan-cache-809e9e85-full.ser
Cannot translate according to [file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.util/module-auto-deps.xml] because could not find /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/lib/boot.jar
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/modules/org-netbeans-api-progress.jar missing for /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.dialogs; will first try to build /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress
platform/api.progress.init-tasks:
platform/api.progress.default.init:
Updating property file: /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/moduleCluster.properties
platform/api.progress.common-init:
platform/api.progress.basic-init:
platform/api.progress.projectized-common.-release.files:
platform/api.progress.projectized-common.files-init:
platform/api.progress.files-init:
platform/api.progress.jdk-8-check:
Build number : 20220921-5c4f116a1a00
platform/api.progress.set-buildnumber:
platform/api.progress.build-init:
Loading module list from /tmp/nb-scan-cache-809e9e85-full.ser
Cannot translate according to [file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.modules/module-auto-deps.xml, file:/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/openide.util/module-auto-deps.xml] because could not find /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/lib/boot.jar
Distilling /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/build/public-package-jars/org-openide-modules.jar from [/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/lib/org-openide-modules.jar]
platform/api.progress.init:
All tests passed
platform/api.progress.up-to-date:
Deleted 20 out of date files in 0 seconds
Compiling 11 source files to /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/build/classes
Ignoring source, target and bootclasspath as release has been set
platform/api.progress.compile:
platform/api.progress.jar-prep:
Building jar: /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/modules/org-netbeans-api-progress.jar
platform/api.progress.projectized-common.jar:
platform/api.progress.jar:
platform/api.progress.netbeans-extra:
platform/api.progress.javahelp:
platform/api.progress.module-auto-deps:
Copying 1 file to /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/netbeans/platform/config/ModuleAutoDeps
platform/api.progress.projectized-common.release:
platform/api.progress.release:
platform/api.progress.module-xml-regular:
platform/api.progress.module-xml-autoload:
platform/api.progress.module-xml-eager:
platform/api.progress.chmod-executables:
platform/api.progress.verify-class-linkage:
0 file(s) have been successfully validated.
1 file(s) have been successfully validated.
platform/api.progress.jar-ml:
platform/api.progress.netbeans:
Generating Auto Update information for org.netbeans.api.progress
platform/api.progress.test-unit-dep-build:
platform/api.progress.test-qa-functional-dep-build:
platform/api.progress.test-dep-build:
platform/api.progress.test-preinit:
platform/api.progress.test-lib-init:
platform/api.progress.test-init:
platform/api.progress.projectized-common.do-unit-test-build:
Created dir: /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/build/test/unit/classes
Compiling 1 source file to /home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/build/test/unit/classes
Ignoring source, target and bootclasspath as release has been set
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:46: error: cannot find symbol
        ProgressContributor contrib1 = AggregateProgressFactory.createProgressContributor("1");
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:47: error: cannot find symbol
        ProgressContributor contrib2 = AggregateProgressFactory.createProgressContributor("2");
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:48: error: cannot find symbol
        AggregateProgressHandle handle = AggregateProgressFactory.createHandle("fact1", new ProgressContributor[] { contrib1, contrib2}, null, null);
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:52: error: cannot find symbol
        ProgressContributor contrib3 = AggregateProgressFactory.createProgressContributor("3");
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:61: error: cannot find symbol
        ProgressContributor contrib1 = AggregateProgressFactory.createProgressContributor("1");
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:62: error: cannot find symbol
        AggregateProgressHandle handle = AggregateProgressFactory.createHandle("fact1", new ProgressContributor[] { contrib1}, null, null);
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:69: warning: [deprecation] getCurrentProgress() in AggregateProgressHandle has been deprecated
        assertEquals(AggregateProgressHandle.WORKUNITS /2, handle.getCurrentProgress());
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:71: error: cannot find symbol
        ProgressContributor contrib2 = AggregateProgressFactory.createProgressContributor("2");
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:75: warning: [deprecation] getCurrentProgress() in AggregateProgressHandle has been deprecated
        assertEquals(AggregateProgressHandle.WORKUNITS /4 * 3, handle.getCurrentProgress());
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:77: error: cannot find symbol
        ProgressContributor contrib3 = AggregateProgressFactory.createProgressContributor("3");
  symbol:   variable AggregateProgressFactory
  location: class AggregateProgressHandleTest
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/platform/api.progress/test/unit/src/org/netbeans/api/progress/aggregate/AggregateProgressHandleTest.java:84: warning: [deprecation] getCurrentProgress() in AggregateProgressHandle has been deprecated
                     handle.getCurrentProgress());
8 errors
3 warnings
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/templates/projectized.xml:153: The following error occurred while executing this line:
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/templates/projectized.xml:153: The following error occurred while executing this line:
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/templates/projectized.xml:153: The following error occurred while executing this line:
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/templates/common.xml:660: The following error occurred while executing this line:
/home/PROGRAMAS/Repositorios_Git/java/Netbeans/netbeans/nbbuild/templates/common.xml:638: Compile failed; see the compiler error output for details.
BUILD FAILED (total time: 0 seconds)

**Comment by mbien** [comment, marked as answer]
1) build netbeans once via command line (https://github.com/apache/netbeans#building-netbeans)
2) incremental updates should work fine if you open the modules you are interested in and re-build them via NB
3) they work like regular projects, you can run/test/debug them. Run will start a dev version of NB.

if you pull from master often or switch branches you will have to do a full clean rebuild periodically.

**Reply by moacirrf** [reply]
Sorry for late response. 
Thanks i will try that now.
